TGeoTessellated volumes read from GDML impossible to be closed (6.40.00 regression)

### Check duplicate issues.

- [x] Checked for duplicates

### Description

In https://github.com/root-project/root/pull/21409, the functionality of `TGeoTessellated::AddFacet` was modified to remove the automatic `CloseShape(check=true)` when the last facet is added.

In https://github.com/root-project/root/pull/21409 also, an early return was added to `CloseShape` (if `fIsClosed && fBVH`) that prevents it from being run multiple times (since `fIsClosed` is set and `BuildBVH` called unconditionally).

Finally (and pre-existing), `TGDMLParse::Tessellated` [always ends](https://github.com/root-project/root/blob/9d249c3554c45b7978b87a80a8f1d31530e0828f/geom/gdml/src/TGDMLParse.cxx#L4439) with `CloseShape(false)`.

These three aspects conspire to **prevent the closure of tessellated volumes** when entered via `TGDMLParse`. The `CloseShape(false)` in `TGDMLParse` essentially locks the default non-closed value since `CheckClosure` is not called. Later `CloseShape(true)` calls have no effect (and worse they just silently return without action).

### Reproducer

Macro (see also zipped/attached below):
```cpp
#include "TGeoManager.h"
#include "TGeoTessellated.h"

void tessellated_closeshape()
{
   // Load a minimal GDML containing a closed tetrahedron (4 triangular faces).
   // TGDMLParse::Tessellated() calls CloseShape(false) after parsing, leaving
   // fClosedBody=false regardless of ROOT version.
   TGeoManager::Import("tetrahedron.gdml");

   auto *tess = dynamic_cast<TGeoTessellated *>(
      gGeoManager->GetListOfShapes()->FindObject("Tetrahedron"));

   if (!tess) {
      printf("[ERROR] Tetrahedron shape not found in tetrahedron.gdml\n");
      return;
   }

   // Call CloseShape(true) to trigger CheckClosure() on the loaded shape.
   // REGRESSION: ROOT v6.40 added `if (fIsClosed && fBVH) return` which blocks
   // this call after CloseShape(false) was already invoked by the GDML parser.
   tess->CloseShape(true);

   bool ok = tess->IsClosedBody();
   printf("[%s] CloseShape(true) after GDML load -> IsClosedBody = %d (expected 1)\n",
          ok ? "PASS" : "FAIL", ok);
}
```

Minimal input gdml file (see also zipped/attached below):
```xml
<?xml version="1.0"?>
<gdml xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
      xsi:noNamespaceSchemaLocation="http://service-spi.web.cern.ch/service-spi/app/releases/GDML/schema/gdml.xsd">
  <define>
    <position name="v0" x="0" y="0" z="0"  unit="cm"/>
    <position name="v1" x="10" y="0" z="0" unit="cm"/>
    <position name="v2" x="5" y="10" z="0" unit="cm"/>
    <position name="v3" x="5" y="5" z="10" unit="cm"/>
  </define>
  <materials>
    <material name="Vacuum" Z="1">
      <D unit="g/cm3" value="1e-25"/>
      <atom unit="g/mole" value="1.008"/>
    </material>
  </materials>
  <solids>
    <tessellated name="Tetrahedron">
      <triangular vertex1="v0" vertex2="v2" vertex3="v1" type="ABSOLUTE"/>
      <triangular vertex1="v0" vertex2="v1" vertex3="v3" type="ABSOLUTE"/>
      <triangular vertex1="v1" vertex2="v2" vertex3="v3" type="ABSOLUTE"/>
      <triangular vertex1="v0" vertex2="v3" vertex3="v2" type="ABSOLUTE"/>
    </tessellated>
  </solids>
  <structure>
    <volume name="World">
      <materialref ref="Vacuum"/>
      <solidref ref="Tetrahedron"/>
    </volume>
  </structure>
  <setup name="Default" version="1.0">
    <world ref="World"/>
  </setup>
</gdml>
```

On ROOT 6.40.00 this produces
```
$ root -l -q tessellated_closeshape.C
Processing tessellated_closeshape.C...
Info in <TGeoManager::Import>: Reading geometry from file: tetrahedron.gdml
Info in <TGeoManager::TGeoManager>: Geometry GDMLImport, Geometry imported from GDML created
Info in <TGeoManager::SetTopVolume>: Top volume is World. Master volume is World
Info in <TGeoNavigator::BuildCache>: --- Maximum geometry depth set to 100
Info in <TGeoManager::CheckGeometry>: Fixing runtime shapes...
Info in <TGeoManager::CheckGeometry>: ...Nothing to fix
Info in <TGeoManager::CloseGeometry>: Counting nodes...
Info in <TGeoManager::Voxelize>: Voxelizing...
Info in <TGeoManager::CloseGeometry>: Building cache...
Info in <TGeoManager::CountLevels>: max level = 1, max placements = 0
Info in <TGeoManager::CloseGeometry>: 1 nodes/ 1 volume UID's in Geometry imported from GDML
Info in <TGeoManager::CloseGeometry>: ----------------modeler ready----------------
[FAIL] CloseShape(true) after GDML load -> IsClosedBody = 0 (expected 1)
```

On ROOT 6.38.00 this produces:
```
$ root -l -q tessellated_closeshape.C

Processing tessellated_closeshape.C...
Info in <TGeoManager::Import>: Reading geometry from file: tetrahedron.gdml
Info in <TGeoManager::TGeoManager>: Geometry GDMLImport, Geometry imported from GDML created
Info in <TGeoManager::SetTopVolume>: Top volume is World. Master volume is World
Info in <TGeoNavigator::BuildCache>: --- Maximum geometry depth set to 100
Info in <TGeoManager::CheckGeometry>: Fixing runtime shapes...
Info in <TGeoManager::CheckGeometry>: ...Nothing to fix
Info in <TGeoManager::CloseGeometry>: Counting nodes...
Info in <TGeoManager::Voxelize>: Voxelizing...
Info in <TGeoManager::CloseGeometry>: Building cache...
Info in <TGeoManager::CountLevels>: max level = 1, max placements = 0
Info in <TGeoManager::CloseGeometry>: 1 nodes/ 1 volume UID's in Geometry imported from GDML
Info in <TGeoManager::CloseGeometry>: ----------------modeler ready----------------
[PASS] CloseShape(true) after GDML load -> IsClosedBody = 1 (expected 1)
```

[tessellated_closeshape.C.zip](https://github.com/user-attachments/files/28180092/tessellated_closeshape.C.zip)
[tetrahedron.gdml.zip](https://github.com/user-attachments/files/28180064/tetrahedron.gdml.zip)

### ROOT version

```
   -------------------------------------------------------------------------------------------------------
  | Welcome to ROOT 6.40.00                                                             https://root.cern |
  | (c) 1995-2025, The ROOT Team; conception: R. Brun, F. Rademakers                                      |
  | Built for linuxx8664gcc on May 23 2026, 20:08:51                                                      |
  | From tags/6-40-00@6-40-00                                                                             |
  | With Ubuntu clang version 23.0.0 (++20260517084559+5f2bedca745d-1~exp1~20260517204734.778) std202002  |
  | Try '.help'/'.?', '.demo', '.license', '.credits', '.quit'/'.q'                                       |
   -------------------------------------------------------------------------------------------------------
```

### Installation method

Spack (develop)

### Operating system

Linux Ubuntu 26.04

### Additional context

A minimal fix would likely be to turn TGDMLParse tessellated solids into closed solids by default, i.e. `tsl->CloseShape(true, true, false)` (aligning with the main use case), but in a sense that just switches to the different default and may affect people who are using gdml imports otherwise.

Introducing early returns without warning for functions like CloseShape should be avoided.

The more fundamental fix is probably to revisit the PR that introduced this regression and disentangle the early return to avoid closure checks from the attempted avoidance of BVH rebuilds.

--- a/geom/geom/src/TGeoTessellated.cxx
+++ b/geom/geom/src/TGeoTessellated.cxx
@@ -328,38 +328,36 @@ bool TGeoTessellated::FacetCheck(int ifacet) const
 
 void TGeoTessellated::CloseShape(bool check, bool fixFlipped, bool verbose)
 {
-   if (fIsClosed && fBVH) {
+   const bool initialized = fIsClosed && fBVH;
+   if (initialized && !check) {
       return;
    }
-   // Compute bounding box
-   fDefined = true;
-   fNvert = fVertices.size();
-   fNfacets = fFacets.size();
-   ComputeBBox();
 
-   BuildBVH();
-   if (fOutwardNormals.size() == 0) {
-      CalculateNormals();
-   } else {
-      // short check if the normal container is of correct size
-      if (fOutwardNormals.size() != fFacets.size()) {
-         std::cerr << "Inconsistency in normal container";
-      }
+   if (!initialized) {
+      // Compute bounding box
+      fDefined = true;
+      fNvert = fVertices.size();
+      fNfacets = fFacets.size();
+      ComputeBBox();
+
+      BuildBVH();
+      fIsClosed = true;
+
+      // Cleanup the vertex map
+      std::multimap<long, int>().swap(fVerticesMap);
    }
-   fIsClosed = true;
-
-   // Cleanup the vertex map
-   std::multimap<long, int>().swap(fVerticesMap);
 
    if (fVertices.size() > 0) {
-      if (!check)
-         return;
+      if (check) {
+         // Check facets
+         for (auto i = 0; i < fNfacets; ++i)
+            FacetCheck(i);
 
-      // Check facets
-      for (auto i = 0; i < fNfacets; ++i)
-         FacetCheck(i);
+         fClosedBody = CheckClosure(fixFlipped, verbose);
+      }
 
-      fClosedBody = CheckClosure(fixFlipped, verbose);
+      if (fOutwardNormals.size() != fFacets.size())
+         CalculateNormals();
    }
 }
 
@@ -422,6 +420,8 @@ bool TGeoTessellated::CheckClosure(bool fixFlipped, bool verbose)
       }
       if (nfixed && verbose)
          Info("Check", "Automatically flipped %d facets to match first defined facet", nfixed);
+      if (nfixed && !fOutwardNormals.empty())
+         CalculateNormals();
    }
    delete[] nn;
    delete[] flipped;
@@ -1277,8 +1277,7 @@ inline Double_t TGeoTessellated::SafetyKernel(const Double_t *point, bool in, in
             const auto object_id = mybvh->prim_ids[p_id];
 
             const auto &facet = fFacets[object_id];
-            auto thissafetySQ =
-               pointFacetDistSq(Vec3f<float>(point[0], point[1], point[2]), facet, fVertices);
+            auto thissafetySQ = pointFacetDistSq(Vec3f<float>(point[0], point[1], point[2]), facet, fVertices);
 
             if (thissafetySQ < smallest_safety_sq) {
                smallest_safety_sq = thissafetySQ;

--- a/geom/test/test_tessellated.cxx
+++ b/geom/test/test_tessellated.cxx
@@ -96,7 +96,38 @@ CreateTrdLikeTessellated_Triangles(const char *name, double x1, double x2, doubl
    return tsl;
 }
 
+void AddClosedTetrahedronFacets(TGeoTessellated &tsl)
+{
+   // Closed tetrahedron using the same facet winding as the GDML reproducer
+   // for ROOT issue #22395.
+   const Vtx v0(0, 0, 0);
+   const Vtx v1(10, 0, 0);
+   const Vtx v2(5, 10, 0);
+   const Vtx v3(5, 5, 10);
+
+   EXPECT_TRUE(tsl.AddFacet(v0, v2, v1));
+   EXPECT_TRUE(tsl.AddFacet(v0, v1, v3));
+   EXPECT_TRUE(tsl.AddFacet(v1, v2, v3));
+   EXPECT_TRUE(tsl.AddFacet(v0, v3, v2));
+}
+
 } // namespace
+
+TEST(TGeoTessellated, CloseShapeCanCheckAfterUncheckedClose)
+{
+   // TGDMLParse finalizes tessellated solids with CloseShape(false). This must
+   // initialize the shape without preventing a later explicit CloseShape(true)
+   // from running closure validation and setting the closed-body state.
+   TGeoTessellated tsl("Closed Tetrahedron");
+   AddClosedTetrahedronFacets(tsl);
+
+   tsl.CloseShape(/*check=*/false);
+   EXPECT_TRUE(tsl.IsDefined());
+   EXPECT_FALSE(tsl.IsClosedBody());
+
+   tsl.CloseShape(/*check=*/true, /*fixFlipped=*/true, /*verbose=*/false);
+   EXPECT_TRUE(tsl.IsClosedBody());
+}
 
 TEST(TGeoTessellated, TrdLike_CoreNavigation)
 {

